### ISSUE TITLE
Server shutdown finalizer crashes

### ISSUE THREAD

**Issue by Byrth** [JuliaWeb/HTTP.jl#697]
Julia 1.5.4
HTTP.jl 0.9.5
MbedTLS.jl 1.0.3 - I am not using SSL on this branch.

I sporadically get two coredump-inspiring errors (maybe 10% / 10% of the time?) when running my tests. I have observed:
* Signal 6: `double free or corruption (!prev)`
* Signal 11: segfault

Both kinds of error involve jl_gc_run_finalizers and occur after it prints my final test when, I presume, the server is being shut down as the process exits.

I have been using HTTP.jl for months, incrementing both it and Julia periodically and I am pretty sure this problem did not exist with HTTP.jl v0.9.1 and Julia 1.5.3. Looking at the diff between 0.9.5 and 0.9.1, I bet it involves the `finally` handling of the `Base.Semaphore` added to src/Servers.jl.

I can extract the core dumps from my dev container if you want, but I'm not sure it's going to be important. Here is what a sig6 printed to the logs:
~~~
*** Error in `/root/julia/bin/julia': double free or corruption (!prev): 0x0000000002e07a40 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7f474)[0x7ff0e320f474]
/lib64/libc.so.6(+0x84770)[0x7ff0e3214770]
/lib64/libc.so.6(realloc+0x1d2)[0x7ff0e3215e12]
/root/julia/bin/../lib/libjulia.so.1(+0x1bb378)[0x7ff0e3d41378]
/root/julia/bin/../lib/libjulia.so.1(+0xfcda0)[0x7ff0e3c82da0]
/root/julia/bin/../lib/libjulia.so.1(+0xfd4a0)[0x7ff0e3c834a0]
/root/julia/bin/../lib/libjulia.so.1(jl_gc_run_pending_finalizers+0x50)[0x7ff0e3c84400]
/root/julia/lib/julia/sys.so(+0xba0d80)[0x7ff0c7535d80]
/root/julia/lib/julia/sys.so(+0xbd34f5)[0x7ff0c75684f5]
[0x7ff0a00c162b]
[0x7ff0a0129f34]
[0x7ff0a012a1d0]
[0x7ff0a012a3d5]
/root/julia/bin/../lib/libjulia.so.1(jl_apply_generic+0x6db)[0x7ff0e3c3920b]
/root/julia/bin/../lib/libjulia.so.1(+0xd1b0f)[0x7ff0e3c57b0f]
======= Memory map: ========
00400000-00402000 r-xp 00000000 fd:03 1643271155                         /root/julia/bin/julia
00602000-00603000 r--p 00002000 fd:03 1643271155                         /root/julia/bin/julia
00603000-00604000 rw-p 00003000 fd:03 1643271155                         /root/julia/bin/julia
00bee000-04789000 rw-p 00000000 00:00 0                                  [heap]
7ff01e9f6000-7ff01fff8000 rw-p 00000000 00:00 0 
7ff01fff8000-7ff028021000 rw-p 00000000 00:00 0 
7ff028021000-7ff02c000000 ---p 00000000 00:00 0 
7ff02c000000-7ff02c008000 ---p 00000000 00:00 0 
7ff02c008000-7ff02c400000 rw-p 00000000 00:00 0 
7ff02c400000-7ff02c408000 ---p 00000000 00:00 0 
7ff02c408000-7ff02c800000 rw-p 00000000 00:00 0 
7ff02c800000-7ff02c808000 ---p 00000000 00:00 0 
7ff02c808000-7ff02cc00000 rw-p 00000000 00:00 0 
7ff02cc00000-7ff02cc08000 ---p 00000000 00:00 0 
7ff02cc08000-7ff02d000000 rw-p 00000000 00:00 0 
7ff02d000000-7ff02d008000 ---p 00000000 00:00 0 
7ff02d008000-7ff02d400000 rw-p 00000000 00:00 0 
7ff02d400000-7ff02d408000 ---p 00000000 00:00 0 
7ff02d408000-7ff02d800000 rw-p 00000000 00:00 0 
7ff02d800000-7ff02d808000 ---p 00000000 00:00 0 
7ff02d808000-7ff02dc00000 rw-p 00000000 00:00 0 
7ff02dc00000-7ff02dc08000 ---p 00000000 00:00 0 
7ff02dc08000-7ff02e000000 rw-p 00000000 00:00 0 
7ff02e000000-7ff02e008000 ---p 00000000 00:00 0 
7ff02e008000-7ff02e400000 rw-p 00000000 00:00 0 
7ff02e400000-7ff02e408000 ---p 00000000 00:00 0 
7ff02e408000-7ff02e800000 rw-p 00000000 00:00 0 
7ff02e800000-7ff02e808000 ---p 00000000 00:00 0 
7ff02e808000-7ff02ec00000 rw-p 00000000 00:00 0 
7ff02ec00000-7ff02ec08000 ---p 00000000 00:00 0 
7ff02ec08000-7ff02f000000 rw-p 00000000 00:00 0 
7ff02f000000-7ff02f008000 ---p 00000000 00:00 0 
7ff02f008000-7ff02f400000 rw-p 00000000 00:00 0 
7ff02f400000-7ff02f408000 ---p 00000000 00:00 0 
7ff02f408000-7ff02f800000 rw-p 00000000 00:00 0 
7ff02f800000-7ff02f808000 ---p 00000000 00:00 0 
7ff02f808000-7ff02fc00000 rw-p 00000000 00:00 0 
7ff02fc00000-7ff02fc08000 ---p 00000000 00:00 0 
7ff02fc08000-7ff030000000 rw-p 00000000 00:00 0 
7ff030000000-7ff0306e3000 rw-p 00000000 00:00 0 
7ff0306e3000-7ff034000000 ---p 00000000 00:00 0 
7ff034057000-7ff035058000 rw-p 00000000 00:00 0 
7ff035058000-7ff035060000 ---p 00000000 00:00 0 
7ff035060000-7ff035458000 rw-p 00000000 00:00 0 
7ff035458000-7ff035460000 ---p 00000000 00:00 0 
7ff035460000-7ff035858000 rw-p 00000000 00:00 0 
7ff035858000-7ff035860000 ---p 00000000 00:00 0 
7ff035860000-7ff035c58000 rw-p 00000000 00:00 0 
7ff035c58000-7ff037ffc000 r--p 00000000 fd:03 554496530                  /root/julia/lib/libjulia.so.1.5
7ff037ffc000-7ff03c021000 rw-p 00000000 00:00 0 
7ff03c021000-7ff040000000 ---p 00000000 00:00 0 
7ff0400b2000-7ff0402b2000 rw-p 00000000 00:00 0 
7ff0402b2000-7ff0403b2000 r-xs 00300000 fd:03 1821906790                 /tmp/tmpfcAX3M0 (deleted)
7ff0403b2000-7ff0403ba000 ---p 00000000 00:00 0 
7ff0403ba000-7ff0407b2000 rw-p 00000000 00:00 0 
7ff0407b3000-7ff0407bb000 ---p 00000000 00:00 0 
7ff0407bb000-7ff040bb3000 rw-p 00000000 00:00 0 
7ff040bb3000-7ff040bbf000 r-xp 00000000 fd:03 553419464                  /usr/lib64/libnss_files-2.17.so
7ff040bbf000-7ff040dbe000 ---p 0000c000 fd:03 553419464                  /usr/lib64/libnss_files-2.17.so
7ff040dbe000-7ff040dbf000 r--p 0000b000 fd:03 553419464                  /usr/lib64/libnss_files-2.17.so
7ff040dbf000-7ff040dc0000 rw-p 0000c000 fd:03 553419464                  /usr/lib64/libnss_files-2.17.so
7ff040dc0000-7ff040dc6000 rw-p 00000000 00:00 0 
7ff040dc6000-7ff040ec6000 r-xs 00200000 fd:03 1821906790                 /tmp/tmpfcAX3M0 (deleted)
7ff040ec6000-7ff040ece000 ---p 00000000 00:00 0 
7ff040ece000-7ff0412c6000 rw-p 00000000 00:00 0 
7ff0412c6000-7ff0412ce000 ---p 00000000 00:00 0 
7ff0412ce000-7ff0416c6000 rw-p 00000000 00:00 0 
7ff0416c6000-7ff0416ce000 ---p 00000000 00:00 0 
7ff0416ce000-7ff041ac6000 rw-p 00000000 00:00 0 
7ff041ac6000-7ff041ace000 ---p 00000000 00:00 0 
7ff041ace000-7ff041ec6000 rw-p 00000000 00:00 0 
7ff041ec7000-7ff0420c7000 rw-p 00000000 00:00 0 
7ff0420c7000-7ff0420cf000 ---p 00000000 00:00 0 
7ff0420cf000-7ff0425c7000 rw-p 00000000 00:00 0 
7ff0425c7000-7ff0425f4000 r-xp 00000000 fd:03 1074767748                 /root/julia/lib/julia/libopenlibm.so.3.0
7ff0425f4000-7ff0427f4000 ---p 0002d000 fd:03 1074767748                 /root/julia/lib/julia/libopenlibm.so.3.0
7ff0427f4000-7ff0427f6000 rw-p 0002d000 fd:03 1074767748                 /root/julia/lib/julia/libopenlibm.so.3.0
7ff0427f6000-7ff0467fa000 rw-p 00000000 00:00 0 
7ff0467fa000-7ff0477fc000 rw-p 00000000 00:00 0 
7ff0477fc000-7ff04b800000 rw-p 00000000 00:00 0 
7ff04b800000-7ff04b808000 ---p 00000000 00:00 0 
7ff04b808000-7ff04bc00000 rw-p 00000000 00:00 0 
7ff04bc00000-7ff04bc08000 ---p 00000000 00:00 0 
7ff04bc08000-7ff04c000000 rw-p 00000000 00:00 0 
7ff04c000000-7ff04c021000 rw-p 00000000 00:00 0 
7ff04c021000-7ff050000000 ---p 00000000 00:00 0 
7ff0500ff000-7ff0501ff000 rw-p 00000000 00:00 0 
7ff0501ff000-7ff050207000 ---p 00000000 00:00 0 
7ff050207000-7ff0505ff000 rw-p 00000000 00:00 0 
7ff0505ff000-7ff050607000 ---p 00000000 00:00 0 
7ff050607000-7ff0509ff000 rw-p 00000000 00:00 0 
7ff0509ff000-7ff050a07000 ---p 00000000 00:00 0 
7ff050a07000-7ff050dff000 rw-p 00000000 00:00 0 
7ff050dff000-7ff050e07000 ---p 00000000 00:00 0 
7ff050e07000-7ff0511ff000 rw-p 00000000 00:00 0 
7ff0511ff000-7ff051207000 ---p 00000000 00:00 0 
7ff051207000-7ff0515ff000 rw-p 00000000 00:00 0 
7ff0515ff000-7ff051607000 ---p 00000000 00:00 0 
7ff051607000-7ff0519ff000 rw-p 00000000 00:00 0 
7ff0519ff000-7ff051a07000 ---p 00000000 00:00 0 
7ff051a07000-7ff051dff000 rw-p 00000000 00:00 0 
7ff051dff000-7ff051e07000 ---p 00000000 00:00 0 
7ff051e07000-7ff0521ff000 rw-p 00000000 00:00 0 
7ff0521ff000-7ff052207000 ---p 00000000 00:00 0 
7ff052207000-7ff0525ff000 rw-p 00000000 00:00 0 
7ff0525ff000-7ff052607000 ---p 00000000 00:00 0 
7ff052607000-7ff0529ff000 rw-p 00000000 00:00 0 
7ff0529ff000-7ff052a07000 ---p 00000000 00:00 0 
7ff052a07000-7ff052dff000 rw-p 00000000 00:00 0 
7ff052dff000-7ff052e07000 ---p 00000000 00:00 0 
7ff052e07000-7ff0531ff000 rw-p 00000000 00:00 0 
7ff0531ff000-7ff053207000 ---p 00000000 00:00 0 
7ff053207000-7ff0535ff000 rw-p 00000000 00:00 0 
7ff0535ff000-7ff053607000 ---p 00000000 00:00 0 
7ff053607000-7ff0539ff000 rw-p 00000000 00:00 0 
signal (6): Aborted
in expression starting at none:0
gsignal at /lib64/libc.so.6 (unknown line)
abort at /lib64/libc.so.6 (unknown line)
__libc_message at /lib64/libc.so.6 (unknown line)
malloc_printerr at /lib64/libc.so.6 (unknown line)
_int_realloc at /lib64/libc.so.6 (unknown line)
realloc at /lib64/libc.so.6 (unknown line)
arraylist_grow at /buildworker/worker/package_linux64/build/src/support/arraylist.c:58 [inlined]
arraylist_push at /buildworker/worker/package_linux64/build/src/support/arraylist.c:69
jl_gc_run_finalizers_in_list at /buildworker/worker/package_linux64/build/src/gc.c:357
run_finalizers at /buildworker/worker/package_linux64/build/src/gc.c:394
jl_gc_run_pending_finalizers at /buildworker/worker/package_linux64/build/src/gc.c:404
enable_finalizers at ./gcutils.jl:108 [inlined]
unlock at ./locks-mt.jl:92
unlock at ./condition.jl:74 [inlined]
unlock at ./lock.jl:133
lock at ./lock.jl:182
lock at ./condition.jl:78 [inlined]
writebusy at /root/.julia/packages/HTTP/cxgat/src/ConnectionPool.jl:46 [inlined]
iswritable at /root/.julia/packages/HTTP/cxgat/src/ConnectionPool.jl:228 [inlined]
iswritable at /root/.julia/packages/HTTP/cxgat/src/Streams.jl:68 [inlined]
closewrite at /root/.julia/packages/HTTP/cxgat/src/Streams.jl:120 [inlined]
writebody at /root/.julia/packages/HTTP/cxgat/src/StreamRequest.jl:117
#2 at ./task.jl:356
unknown function (ip: 0x7ff0a012a3d4)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2214 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2398
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1691 [inlined]
start_task at /buildworker/worker/package_linux64/build/src/task.c:721
unknown function (ip: (nil))
Allocations: 242988916 (Pool: 242901426; Big: 87490); GC: 252
ERROR: Package JuliaServer errored during testing (received signal: 6)
~~~

**Comment by quinnj** [comment]
Digging into this a bit, I think we're resolved on HTTP#master. Before the multithreading safety overhaul, there were cases where tasks would get stuck in the request pipeline and I think the finalizers here (plus stacktraces) show we were waiting on those. Closing for now, but feel free to chime in or open a new issue if people see issues on HTTP#master.
